gcp_variant_transforms/libs/partitioning.py

```diff
@@ -182,59 +181,64 @@ def _copy_to_flatten_table(self, output_table_id, cp_query):
+        call_mode = schema_item.get('mode')
+        if call_mode != bigquery_util.TableFieldConstants.MODE_REPEATED:
+          logging.error('Expected REPEATED mode for column `call` but got: %s',
+                        call_mode)
+          raise ValueError('Wrong mode for column `call`: {}'.format(call_mode))
+        schema_item['mode'] = bigquery_util.TableFieldConstants.MODE_NULLABLE
+        # (2) Duplicate sample_id as an independent column to the table
+        sub_items = schema_item.get('fields')
+        sample_id_found = False
+        for sub_item in sub_items:
+          if (sub_item.get('name') ==
+              bigquery_util.ColumnKeyConstants.CALLS_SAMPLE_ID):
+            schema_json.append(sub_item)
+            sample_id_found = True
+            break
+        if not sample_id_found:
+          logging.info('`sample_id` column under `call` column was not found.')
+          raise ValueError(
+              '`sample_id` column under `call` column was not found.')
+      schema_json.append(schema_item)
+    return schema_json
 
   def get_flatten_table_schema(self, schema_file_path):
     # type: (str) -> bool
     """Write the flatten table's schema to the given json file.
 
     This method basically performs the following tasks:
-      * Composes a 'flattening query' based on _schema_table_id table's schema.
-      * Runs the 'flattening query' to read 1 row and writes it to a temp table.
-      * Extracts the schema of the temp table using _BQ_EXTRACT_SCHEMA_COMMAND.
+      * Extract variant table schema using BigQuery API.
+      * Copy all columns without any change except `call` column:
+        * Modify mode from REPEATED TO NULLABLE
+        * Duplicate call.sample_id column as sample_id column (for partitioning)
 
     Args:
       schema_file_path: The json schema will be written to this file.
 
     Returns;
       A bool value indicating if the schema was successfully extracted.
     """
-    temp_table_id = self._create_temp_flatten_table_with_1_row()
-    full_table_id = '{}:{}.{}'.format(
-        self._project_id, self._dataset_id, temp_table_id)
-    bq_command = _BQ_EXTRACT_SCHEMA_COMMAND.format(
-        FULL_TABLE_ID=full_table_id,
-        SCHEMA_FILE_PATH=schema_file_path)
-    result = os.system(bq_command)
-    if result != 0:
-      logging.error('Failed to extract flatten table schema using "%s" command',
-                    bq_command)
-    else:
-      logging.info('Successfully extracted the schema of flatten table.')
-    if bigquery_util.delete_table(full_table_id) == 0:
-      logging.info('Successfully deleted temporary table: %s', full_table_id)
-    else:
-      logging.error('Was not able to delete temporary table: %s', full_table_id)
-    return result == 0
+    full_table_id = '{}.{}.{}'.format(
+        self._project_id, self._dataset_id, self._schema_table_id)
+    try:
+      variant_table = self._client.get_table(full_table_id)
```

Comment:
Do we need to get schema from BQ table? Don't we already have it computed at this point?

Reply:
Probably we could get it from BEAM, however, since we are creating sample optimized tables from variant optimized tables I can imagine in future we turn this into an independent functionality. That's why I thought we better not to have dependency on other parts of code.